File: binary-tree-bfs/ZigZagTraversal.java
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/?envType=study-plan-v2&envId=top-interview-150

import java.util.*;

// Status: passed 19/33 testcases -> good enough for phase 1

/**
- start at root, add the root element to the queue
- initialize a counter
- if current depth is even then 
  - break out if queue is empty
  - pull elements from the queue, add non null children to a stack from right to left, 
  - also add pulled elements to the final output
- if current depth is odd then
  - break out if stack is empty
  - pull/remove elements from the stack, add non null children to queue from left to right, 
  - also add pulled elements to the final output
 */


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {


        int depth = 0;
        List<List<Integer>> outputList = new ArrayList<>();
        Queue<TreeNode> levelQueue = new LinkedList<TreeNode>();
        Stack<TreeNode> levelStack = new Stack<TreeNode>();
        levelQueue.add(root);

        if( root == null ) {
            return outputList;
        }

        while(true) {

            if(depth % 2 == 0) {
                if(levelQueue.size() == 0) {
                    break;
                }

                List<Integer> currentValues = new ArrayList<>();

                while(levelQueue.size()!= 0) {
                    TreeNode currentNode = levelQueue.poll();

                    if(currentNode.left != null) {
                        levelStack.add(currentNode.left);
                    }

                    if(currentNode.right != null) {
                        levelStack.add(currentNode.right);
                    }

                    currentValues.add(currentNode.val);
                }

                outputList.add(currentValues);
            } else {
                if(levelStack.size() == 0) {
                    break;
                }

                List<Integer> currentValues = new ArrayList<>();

                while(levelStack.size() != 0) {
                    TreeNode currentNode = levelStack.pop();

                    if(currentNode.left != null) {
                        levelQueue.add(currentNode.left);
                    }

                    if(currentNode.right != null) {
                        levelQueue.add(currentNode.right);
                    }

                    currentValues.add(currentNode.val);
                }

                outputList.add(currentValues);
            }

            depth++;
        }
        

        return outputList;
    }
}

File: binary-tree-bfs/ZigZagTraversal.py
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/?envType=study-plan-v2&envId=top-interview-150

from typing import List, Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
        
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if not root:
            return []
        
        node_queue = deque([root])
        output = []
        reverse = False

        while node_queue:
            node_queue_size = len(node_queue)
            temp_level_arr = []

            for i in range(node_queue_size):
                if reverse: 
                    print("picking node from end")
                    node = node_queue.pop()
                else:
                    print("picking node from beginning")
                    node = node_queue.popleft()
                print("popped out: ", node.val)

                if node.left:
                        if reverse:
                            node_queue.appendleft(node.left)    
                        else:
                            node_queue.append(node.left)

                if node.right: 
                        if reverse:
                            node_queue.appendleft(node.right)    
                        else:
                            node_queue.append(node.right)

                # print(node_queue)
                temp_level_arr.append(node.val)

            output.append(temp_level_arr)
            reverse = not reverse

        return output